### DISCUSSION TITLE
Are conditional type hints currently possible?

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1251] [Q&A] [answered]
Consider the case when we have a table of pairs of types (cf. https://github.com/pandas-dev/pandas-stubs/issues/262)

| TypeA        | TypeB                 |
|--------------|-----------------------|
| `np.int32`   | `pandas.Int32Dtype`   |
| `np.int64`   | `pandas.Int64Dtype`   |
| `np.float32` | `pandas.Float32Dtype` |
| `np.float64` | `pandas.Float64Dtype` |

And I have a functon that takes as input an object of the form `list[NumpyScalar]`, returns a `list[PandasScalar]` of the corresponding type from the Lookup table.

```python
def convert(object: list[NumpyScalar]) -> list[PandasScalar]:
```

How can one type hint this properly without writing a million `@overload`s?

**Comment by hmc-cs-mdrissi** [comment, marked as answer]
No overload is only way today.

You need a type level map operation which has briefly appeared in typing sig and while useful would likely be a complex feature and definitely require a pep. I think numpy dtype conversion/promotion rules is best use case for this.
